### PR TITLE
#210: Datasource packetfabric_port_device_info

### LINKED ISSUE #210: data source packetfabric_port_device_info
Get port device info

data source `packetfabric_port_device_info` with port_circuit_id as user input

GET [/v2/ports/{port_circuit_id}/device-info](https://docs.packetfabric.com/api/v2/redoc/#tag/Ports/operation/get_port_device_info)

Checklist:
- [ ] tested locally
- [ ] added/updated acceptance tests
- [ ] add examples (in example folder)
- [ ] generate docs

### PR DESCRIPTION
**Pull request by pontovinte**
Signed-off-by: Yrineu Rodrigues <yrineu@pontovinte.com>

## Description

Datasource `packetfabric_port_device_info`

Closes #210 

## Checklist

- [x] tested locally
- [x] added/updated acceptance tests
- [x] add examples (in example folder)
- [x] generate docs

### CODE REVIEW

**Comment by medzin** [comment]
Acceptance test added.

**Review by elXaptation — DISMISSED** [review]

File: templates/data-sources/port_device_info.md.tmpl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+# packetfabric_port_device_info (Data Source)
+
+
+
+## Example Usage
+
+{{tffile "examples/data-sources/packetfabric_port_device_info/data-source.tf"}}
+
+<!-- schema generated by tfplugindocs -->
```
Comment (rjouhann) [inline templates/data-sources/port_device_info.md.tmpl:17]: @elXaptation the schema should not be included in the `tmpl` file I think.

File: examples/data-sources/packetfabric_port_device_info/data-source.tf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+data "packetfabric_port_device_info" "port_device_info_1" {
+  provider = packetfabric
+  port_circuit_id = var.pf_port_circuit_id
+}
```
Comment (rjouhann) [inline examples/data-sources/packetfabric_port_device_info/data-source.tf:4]: @elXaptation should run `terraform fmt` on the examples too.

**Review by rjouhann — CHANGES_REQUESTED** [review]

**Review by rjouhann — CHANGES_REQUESTED** [review]
@pontovinte I am getting this error when I try:

```
data.packetfabric_port_device_info.port_device_info_1: Reading...
2022-11-29T17:41:17.754Z [ERROR] provider.terraform-provider-packetfabric: Response contains error diagnostic: diagnostic_severity=ERROR diagnostic_summary="json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64" tf_data_source_type=packetfabric_port_device_info tf_proto_version=5.3 tf_provider_addr=provider @caller=/root/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail= tf_req_id=0ff0b98c-9bab-8117-3a25-28c4f7424a1a tf_rpc=ReadDataSource timestamp=2022-11-29T17:41:17.753Z
2022-11-29T17:41:17.755Z [ERROR] vertex "data.packetfabric_port_device_info.port_device_info_1" error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
2022-11-29T17:41:17.756Z [ERROR] vertex "data.packetfabric_port_device_info.port_device_info_1 (expand)" error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
╷
│ Error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
│ 
│   with data.packetfabric_port_device_info.port_device_info_1,
│   on main.tf line 10, in data "packetfabric_port_device_info" "port_device_info_1":
│   10: data "packetfabric_port_device_info" "port_device_info_1" {
│ 
```

**Comment by pontovinte** [comment]
> @pontovinte I am getting this error when I try:
> 
> ```
> data.packetfabric_port_device_info.port_device_info_1: Reading...
> 2022-11-29T17:41:17.754Z [ERROR] provider.terraform-provider-packetfabric: Response contains error diagnostic: diagnostic_severity=ERROR diagnostic_summary="json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64" tf_data_source_type=packetfabric_port_device_info tf_proto_version=5.3 tf_provider_addr=provider @caller=/root/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail= tf_req_id=0ff0b98c-9bab-8117-3a25-28c4f7424a1a tf_rpc=ReadDataSource timestamp=2022-11-29T17:41:17.753Z
> 2022-11-29T17:41:17.755Z [ERROR] vertex "data.packetfabric_port_device_info.port_device_info_1" error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
> 2022-11-29T17:41:17.756Z [ERROR] vertex "data.packetfabric_port_device_info.port_device_info_1 (expand)" error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
> ╷
> │ Error: json: cannot unmarshal string into Go struct field OpticsDiagnosticsLaneValues.optics_diagnostics_lane_values.lane_index of type float64
> │ 
> │   with data.packetfabric_port_device_info.port_device_info_1,
> │   on main.tf line 10, in data "packetfabric_port_device_info" "port_device_info_1":
> │   10: data "packetfabric_port_device_info" "port_device_info_1" {
> │ 
> ```
> 
> I did the equivalent GET on `{{url}}/v2/ports/PF-AP-WDC1-1726464/device-info`
> 
> ```
> {
>     "adjacent_router": null,
>     "device_id": 836,
>     "device_name": "cfr7.wdc1",
>     "device_make": "cisco",
>     "admin_status": "up",
>     "oper_status": "down",
>     "auto_negotiation": false,
>     "iface_name": "te0/0/0/30",
>     "speed": "10Gbps",
>     "site_id": 344,
>     "optics_diagnostics_lane_values": [
>         {
>             "tx_power_dbm": -3.01,
>             "tx_power": 0.5,
>             "lane_index": "0",
>             "rx_power": 0.0,
>             "rx_power_dbm": -40.0,
>             "bias_current": 31.22,
>             "tx_status": "",
>             "rx_status": "low_alarm"
>         }
>     ],
>     "polltime": null,
>     "time_flapped": "-1",
>     "traffic_rx_bps": 0,
>     "traffic_rx_bytes": 0,
>     "traffic_rx_ipv6_bytes": 0,
>     "traffic_rx_ipv6_packets": 0,
>     "traffic_rx_packets": 0,
>     "traffic_rx_pps": 0,
>     "traffic_tx_bps": 0,
>     "traffic_tx_bytes": 0,
>     "traffic_tx_ipv6_bytes": 0,
>     "traffic_tx_ipv6_packets": 0,
>     "traffic_tx_packets": 0,
>     "traffic_tx_pps": 0,
>     "wiring_media": "LR",
>     "wiring_module": "unknown",
>     "wiring_panel": "DC5:051305:0000:1045813",
>     "wiring_position": "49",
>     "wiring_reach": "10km",
>     "wiring_type": "LC/SMF",
>     "lag_speed": 0,
>     "device_can_lag": true
> }
> ```

Is there any inconsistency with the [API](https://docs.packetfabric.com/api/v2/redoc/#tag/Ports/operation/get_port_device_info) or there was any recent changes that affected this API? I did test it locally without experiencing that.

**Comment by rjouhann** [comment]
> 
> Is there any inconsistency with the [API](https://docs.packetfabric.com/api/v2/redoc/#tag/Ports/operation/get_port_device_info) or there was any recent changes that affected this API? I did test it locally without experiencing that.

Yes, it looks like the API spec isn't up to date. I will get it updated. In the meantime, could you use the above payload to add the missing attributes?

I identified the following ones missing:

- adjacent_router
- device_id
- site_id
- optics_diagnostics_lane_values[].tx_status
- optics_diagnostics_lane_values[].tx_status
- lag_speed
- device_can_lag

**Comment by pontovinte** [comment]
I just sent a new patch adding the `RX` and `TX` status.

I'm testing against Prod env and I'm getting the following results:

```
Outputs:

packetfabric_port_device_info = {
  "adjacent_router" = tostring(null)
  "admin_status" = "up"
  "auto_negotiation" = true
  "device_can_lag" = false
  "device_id" = 216
  "device_make" = "juniper"
  "device_name" = "efr1.pdx1"
  "id" = "PF-AP-PDX1-1750000"
  "iface_name" = "ge-0/0/17"
  "lag_speed" = 0
  "oper_status" = "down"
  "optics_diagnostics_lane_bias_current" = 17.45
  "optics_diagnostics_lane_index" = 0
  "optics_diagnostics_lane_rx_power" = 0
  "optics_diagnostics_lane_rx_power_dbm" = -50
  "optics_diagnostics_lane_rx_status" = "low_alarm"
  "optics_diagnostics_lane_tx_power" = 0.238
  "optics_diagnostics_lane_tx_power_dbm" = -6.23
  "optics_diagnostics_lane_tx_status" = ""
  "polltime" = tonumber(null)
  "port_circuit_id" = "PF-AP-PDX1-1750000"
  "site_id" = 388
  "speed" = "1Gbps"
  "time_flapped" = "91130568"
  "traffic_rx_bps" = 0
  "traffic_rx_bytes" = 0
  "traffic_rx_ipv6_bytes" = 0
  "traffic_rx_ipv6_packets" = 0
  "traffic_rx_packets" = 0
  "traffic_rx_pps" = 0
  "traffic_tx_bps" = 0
  "traffic_tx_bytes" = 0
  "traffic_tx_ipv6_bytes" = 0
  "traffic_tx_ipv6_packets" = 0
  "traffic_tx_packets" = 0
  "traffic_tx_pps" = 0
  "wiring_media" = "LX"
  "wiring_module" = "-"
  "wiring_panel" = "T250.03.07/04."
  "wiring_position" = "83|84"
  "wiring_reach" = "10km"
  "wiring_type" = "SC/SMF"
}

```

**Comment by rjouhann** [comment]
Tested with both PacketFabric and Colt ports and things works perfect. Thanks!

**Review by rjouhann — APPROVED** [review]

**Review by caitlinwheeless — APPROVED** [review]
No other doc changes needed

**Pull request merged**